### ISSUE TITLE
Movement is laggy, possible solution

### ISSUE THREAD

**Issue by maartenslaa** [googlevr/cardboard#193]
It seems like the current solution does not predict rotation values correctly. I have done some digging and found that when I change this (in head_tracker.cc -> void HeadTracker::GetPose):

`predicted_rotation = pose_state.sensor_from_start_rotation;`
to this:
`predicted_rotation = pose_prediction::PredictPose(timestamp_ns, pose_state);`

movement is fast and stable on iOS (iPhone 12 Pro max). I have tried the same on Android (Samsung S8) but this needs another fix since the time scopes of timestamp_ns and the one set in pose_state are different. I changed this (in util.cc -> long GetMonotonicTimeNano):

`clock_gettime(CLOCK_MONOTONIC, &res);`
to this:
`clock_gettime(CLOCK_BOOTTIME, &res);`

which made it work on the Samsung S8 as well. Much better result than the current implementation.

I'm not sure of other consequences yet, does it still do Kalman filtering and bias prediction this way?

**Comment by maartenslaa** [comment]
Same positive result on Huawei P10 Lite, Nokia 5 and iPhone 6s. I think this may solve issues #180 and #163 as wel.

**Comment by Treecrotch** [comment]
Hi! Yeah that sounds like you might be right about these other reported issues all being the same, described as slow/lag/blur/choppy...not as responsive as old sdk or other older example apps. 

Feeling a bit silly here but, if you've indeed fixed it how do i test this in unity? Am i correct in assuming you are referring to files in the native sdk, where in unity I would need to first compile a new xr plugin? I recall looking for those files to edit in the recent past but nonsuch files exist as im using the xr plug in

**Comment by maartenslaa** [comment]
You are correct, I am referring to the native SDK. I got it to work on iOS with Unity as well though. There are some files in the native SDK that are used for Unity as well. You will still need to do the above, and change cardboard_xr_unity.cc as well (I haven't tested it for Android so not sure if the change for CLOCK_MONOTONIC to CLOCK_BOOTTIME is required but I think it is):
`static int64_t GetMonotonicTimeNano() {`
    `struct timespec res;`
    `#ifdef __APPLE__`
     `clock_gettime(CLOCK_UPTIME_RAW, &res);`
`#else`
      `clock_gettime(CLOCK_BOOTTIME, &res);`
`#endif`
    `return (res.tv_sec * CardboardApi::CardboardApiImpl::kNanosInSeconds) + res.tv_nsec;`
 ` }`
  You can build the project with Android Studio and replace the sdk.aar file with the one generated.
  
  When I get to testing the Android version of the Unity implementation I will post an SDK for you to test.

**Comment by Treecrotch** [comment]
Thank you. 

In the mean time you mentioned you tested on the Huawei P10 Lite with good results. if u can post that .apk file somewhere i would be happy to test it out on my android devices that were showing rotational lag and report back. (pixel xl4 and moto g5)

**Comment by maartenslaa** [comment]
Here you go: [HelloCardboard test APK](https://aryzoncom-my.sharepoint.com/:u:/g/personal/info_aryzon_com1/EenPFn1ybTtBhtzmBPGo7G0BUNTAa4showKvfkWtVmAp0A?e=dfiddH)

**Comment by maartenslaa** [comment]
This shows the sensor timestamps are in SystemClock.elapsedRealtimeNanos() frame:
https://developer.android.com/reference/android/hardware/SensorEvent.html#timestamp

This shows SystemClock.elapsedRealtimeNanos() actually is clock_gettime(CLOCK_BOOTTIME, &res):
https://android.googlesource.com/platform/system/core/+/master/libutils/SystemClock.cpp

**Comment by Treecrotch** [comment]
Thanks, unfortunately it won't install, stating a generic "you can't install the app on your device"

I do have "installation from unknown sources" enabled and have recently installed other test apks from unknown sources, browser, or random file downloads. For whatever reason this apk seems to not work for me. The file name says its debug version, maybe needs a release version?

If you end up building a release version let me know, otherwise i don't want to take a bunch of your time on it, it sounds like it is the fix though. I was the original poster of the "blur" bug you tagged. But essentially what i meant to say in that bug is its laggy on rotation

**Comment by maartenslaa** [comment]
You mentioned you are using Unity, it only takes replacing one file to make it work in your own project.

Make sure you have version 1.4.1 of the Cardboard SDK installed in the package manager. Then in your project tree go to Packages -> Google Cardboard XR Plugin for Unity -> Runtime -> Android. Replace the sdk.aar file with the one below, but make sure to keep the sdk.meta file otherwise Unity won't use the file while compiling.

[sdk.aar](https://aryzoncom-my.sharepoint.com/:u:/g/personal/info_aryzon_com1/EcsCePyX9NdEm2dD7Z8L9kUBykHc8lOYc7tGfK82abSLnQ?e=skgVEu)

**Comment by Treecrotch** [comment]
Oh ok great ill give it a shot. I thought i had to build the whole plugin with Android studio

**Comment by Treecrotch** [comment]
I tired it on the pixel xl4, it might be a tad better but still seems a bit jitter compared to other VR apps on the play store or youtube vr. 

overlall the SDK just produces less quality experience compared to older apps

**Comment by Excubitor** [comment]
Have anyone built this for Unity and iOS and have the file to replace? I would like to try this.

**Comment by Excubitor** [comment]
Tried to do the change above and replace the cardboard_api.a file in unity but got the same problem I got when using the old google vr sdk and latest version of iOS, the view is constantly spinning.

**Comment by maartenslaa** [comment]
Did you set clock_gettime(CLOCK_UPTIME_RAW, &res); in cardboard_xr_unity.cc? If not the view's rotation will be all over the place. This is the one I just build. I cannot test stereoscopic view for iOS because I use a Unity version where I cannot set OpenGLES. Should work though:

[cardboard-api.a](https://aryzoncom-my.sharepoint.com/:u:/g/personal/info_aryzon_com1/EcLLuyDKG6lDkRtT5gu7CuwB7XDnOqkISMO7on4Rb4IouA?e=hlts0c)

**Comment by Excubitor** [comment]
Thanks, that worked. I'm using an iPhone pro 12 and I think I see some improvement when turning my head, but I'm guessing there is an update limit on the sensor that limits how good it can work.

**Comment by hyenachi** [comment]
I edited codes as you guided and build it.
I tested  Samsung G8, G10. I feel it is same. 
Motion blur is still remain when head tracking.

**Comment by jballoffet** [comment]
Thanks for reaching out to us with this proposal. After doing a deep code analysis, a fix for this has just been submitted and it will be included in the next release. In case you cannot wait until v1.5.0 is out, you can apply the required changes by yourself. In order to do so, you just need to apply the following patch:

```bash
diff --git a/hellocardboard-android/src/main/jni/util.cc b/hellocardboard-android/src/main/jni/util.cc
index 0da55d8..57aa636 100644
--- a/hellocardboard-android/src/main/jni/util.cc
+++ b/hellocardboard-android/src/main/jni/util.cc
@@ -428,9 +428,9 @@ float AngleBetweenVectors(const std::array<float, 4>& vec1,
 
 static constexpr uint64_t kNanosInSeconds = 1000000000;
 
-long GetMonotonicTimeNano() {
+int64_t GetBootTimeNano() {
   struct timespec res;
-  clock_gettime(CLOCK_MONOTONIC, &res);
+  clock_gettime(CLOCK_BOOTTIME, &res);
   return (res.tv_sec * kNanosInSeconds) + res.tv_nsec;
 }
 
diff --git a/hellocardboard-android/src/main/jni/util.h b/hellocardboard-android/src/main/jni/util.h
index 4af8b8f..f857b68 100644
--- a/hellocardboard-android/src/main/jni/util.h
+++ b/hellocardboard-android/src/main/jni/util.h
@@ -93,11 +93,11 @@ float AngleBetweenVectors(const std::array<float, 4>& vec1,
                           const std::array<float, 4>& vec2);
 
 /**
- * Gets monotonic time in nanoseconds.
+ * Gets system boot time in nanoseconds.
  *
- * @return Monotonic time in nanoseconds
+ * @return System boot time in nanoseconds
  */
-long GetMonotonicTimeNano();
+int64_t GetBootTimeNano();
 
 /**
  * Generates a random floating point number between |min| and |max|.
diff --git a/sdk/head_tracker.cc b/sdk/head_tracker.cc
index 22318df..e2221f1 100644
--- a/sdk/head_tracker.cc
+++ b/sdk/head_tracker.cc
@@ -67,13 +67,7 @@ void HeadTracker::GetPose(int64_t timestamp_ns,
                           std::array<float, 4>& out_orientation) const {
   Rotation predicted_rotation;
   const PoseState pose_state = sensor_fusion_->GetLatestPoseState();
-  if (!sensor_fusion_->IsFullyInitialized()) {
-    CARDBOARD_LOGI(
-        "Head Tracker not fully initialized yet. Using pose prediction only.");
-    predicted_rotation = pose_prediction::PredictPose(timestamp_ns, pose_state);
-  } else {
-    predicted_rotation = pose_state.sensor_from_start_rotation;
-  }
+  predicted_rotation = pose_prediction::PredictPose(timestamp_ns, pose_state);
 
   // In order to update our pose as the sensor changes, we begin with the
   // inverse default orientation (the orientation returned by a reset sensor),
@@ -85,16 +79,14 @@ void HeadTracker::GetPose(int64_t timestamp_ns,
   const Rotation sensor_to_display =
       Rotation::FromAxisAndAngle(Vector3(0, 0, 1), M_PI / 2.0);
 
-  const Vector4 q =
+  const Vector4 orientation =
       (sensor_to_display * predicted_rotation * ekf_to_head_tracker)
           .GetQuaternion();
-  Rotation rotation;
-  rotation.SetQuaternion(q);
 
-  out_orientation[0] = static_cast<float>(rotation.GetQuaternion()[0]);
-  out_orientation[1] = static_cast<float>(rotation.GetQuaternion()[1]);
-  out_orientation[2] = static_cast<float>(rotation.GetQuaternion()[2]);
-  out_orientation[3] = static_cast<float>(rotation.GetQuaternion()[3]);
+  out_orientation[0] = static_cast<float>(orientation[0]);
+  out_orientation[1] = static_cast<float>(orientation[1]);
+  out_orientation[2] = static_cast<float>(orientation[2]);
+  out_orientation[3] = static_cast<float>(orientation[3]);
 
   out_position = ApplyNeckModel(out_orientation, 1.0);
 }
diff --git a/sdk/include/cardboard.h b/sdk/include/cardboard.h
index 74a0a67..ed00438 100644
--- a/sdk/include/cardboard.h
+++ b/sdk/include/cardboard.h
@@ -383,6 +383,13 @@ void CardboardHeadTracker_resume(CardboardHeadTracker* head_tracker);
 
 /// Gets the predicted head pose for a given timestamp.
 ///
+/// @details On Android devices, @p timestamp_ns must be in system boot time
+///          (see CLOCK_BOOTTIME in https://linux.die.net/man/2/clock_gettime)
+///          clock (see https://developer.android.com/reference/android/hardware/SensorEvent#timestamp).
+///          On iOS devices, @p timestamp_ns must be in system uptime raw
+///          (see CLOCK_UPTIME_RAW in http://www.manpagez.com/man/3/clock_gettime/)
+///          clock (see https://developer.apple.com/documentation/coremotion/cmlogitem/1615939-timestamp?language=objc).
+///
 /// @pre @p head_tracker Must not be null.
 /// @pre @p position Must not be null.
 /// @pre @p orientation Must not be null.
@@ -391,7 +398,7 @@ void CardboardHeadTracker_resume(CardboardHeadTracker* head_tracker);
 ///
 /// @param[in]      head_tracker            Head tracker object pointer.
 /// @param[in]      timestamp_ns            The timestamp for the pose in
-///     nanoseconds in system monotonic clock.
+///                                         nanoseconds.
 /// @param[out]     position                3 floats for (x, y, z).
 /// @param[out]     orientation             4 floats for quaternion
 void CardboardHeadTracker_getPose(CardboardHeadTracker* head_tracker,
diff --git a/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc b/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
index 0cee973..d5af217 100644
--- a/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
+++ b/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
@@ -271,7 +271,7 @@ class CardboardApi::CardboardApiImpl {
       return;
     }
     CardboardHeadTracker_getPose(head_tracker_.get(),
-                                 CardboardApiImpl::GetMonotonicTimeNano() +
+                                 CardboardApiImpl::GetBootTimeNano() +
                                      kPredictionTimeWithoutVsyncNanos,
                                  position, orientation);
   }
@@ -476,11 +476,15 @@ class CardboardApi::CardboardApiImpl {
     }
   };
 
-  // @brief Computes the monotonic time in nano seconds.
-  // @return The monotonic time count in nano seconds.
-  static int64_t GetMonotonicTimeNano() {
+  // @brief Computes the system boot time in nanoseconds.
+  // @return The system boot time count in nanoseconds.
+  static int64_t GetBootTimeNano() {
     struct timespec res;
-    clock_gettime(CLOCK_MONOTONIC, &res);
+#if defined(__ANDROID__)
+    clock_gettime(CLOCK_BOOTTIME, &res);
+#elif defined(__APPLE__)
+    clock_gettime(CLOCK_UPTIME_RAW, &res);
+#endif
     return (res.tv_sec * CardboardApi::CardboardApiImpl::kNanosInSeconds) +
            res.tv_nsec;
   }
```

**Comment by jeff-rey-unity** [comment]
> Thanks for reaching out to us with this proposal. After doing a deep code analysis, a fix for this has just been submitted and it will be included in the next release. In case you cannot wait until v1.5.0 is out, you can apply the required changes by yourself. In order to do so, you just need to apply the following patch:
> 
> ```shell
> diff --git a/hellocardboard-android/src/main/jni/util.cc b/hellocardboard-android/src/main/jni/util.cc
> index 0da55d8..57aa636 100644
> --- a/hellocardboard-android/src/main/jni/util.cc
> +++ b/hellocardboard-android/src/main/jni/util.cc
> @@ -428,9 +428,9 @@ float AngleBetweenVectors(const std::array<float, 4>& vec1,
>  
>  static constexpr uint64_t kNanosInSeconds = 1000000000;
>  
> -long GetMonotonicTimeNano() {
> +int64_t GetBootTimeNano() {
>    struct timespec res;
> -  clock_gettime(CLOCK_MONOTONIC, &res);
> +  clock_gettime(CLOCK_BOOTTIME, &res);
>    return (res.tv_sec * kNanosInSeconds) + res.tv_nsec;
>  }
>  
> diff --git a/hellocardboard-android/src/main/jni/util.h b/hellocardboard-android/src/main/jni/util.h
> index 4af8b8f..f857b68 100644
> --- a/hellocardboard-android/src/main/jni/util.h
> +++ b/hellocardboard-android/src/main/jni/util.h
> @@ -93,11 +93,11 @@ float AngleBetweenVectors(const std::array<float, 4>& vec1,
>                            const std::array<float, 4>& vec2);
>  
>  /**
> - * Gets monotonic time in nanoseconds.
> + * Gets system boot time in nanoseconds.
>   *
> - * @return Monotonic time in nanoseconds
> + * @return System boot time in nanoseconds
>   */
> -long GetMonotonicTimeNano();
> +int64_t GetBootTimeNano();
>  
>  /**
>   * Generates a random floating point number between |min| and |max|.
> diff --git a/sdk/head_tracker.cc b/sdk/head_tracker.cc
> index 22318df..e2221f1 100644
> --- a/sdk/head_tracker.cc
> +++ b/sdk/head_tracker.cc
> @@ -67,13 +67,7 @@ void HeadTracker::GetPose(int64_t timestamp_ns,
>                            std::array<float, 4>& out_orientation) const {
>    Rotation predicted_rotation;
>    const PoseState pose_state = sensor_fusion_->GetLatestPoseState();
> -  if (!sensor_fusion_->IsFullyInitialized()) {
> -    CARDBOARD_LOGI(
> -        "Head Tracker not fully initialized yet. Using pose prediction only.");
> -    predicted_rotation = pose_prediction::PredictPose(timestamp_ns, pose_state);
> -  } else {
> -    predicted_rotation = pose_state.sensor_from_start_rotation;
> -  }
> +  predicted_rotation = pose_prediction::PredictPose(timestamp_ns, pose_state);
>  
>    // In order to update our pose as the sensor changes, we begin with the
>    // inverse default orientation (the orientation returned by a reset sensor),
> @@ -85,16 +79,14 @@ void HeadTracker::GetPose(int64_t timestamp_ns,
>    const Rotation sensor_to_display =
>        Rotation::FromAxisAndAngle(Vector3(0, 0, 1), M_PI / 2.0);
>  
> -  const Vector4 q =
> +  const Vector4 orientation =
>        (sensor_to_display * predicted_rotation * ekf_to_head_tracker)
>            .GetQuaternion();
> -  Rotation rotation;
> -  rotation.SetQuaternion(q);
>  
> -  out_orientation[0] = static_cast<float>(rotation.GetQuaternion()[0]);
> -  out_orientation[1] = static_cast<float>(rotation.GetQuaternion()[1]);
> -  out_orientation[2] = static_cast<float>(rotation.GetQuaternion()[2]);
> -  out_orientation[3] = static_cast<float>(rotation.GetQuaternion()[3]);
> +  out_orientation[0] = static_cast<float>(orientation[0]);
> +  out_orientation[1] = static_cast<float>(orientation[1]);
> +  out_orientation[2] = static_cast<float>(orientation[2]);
> +  out_orientation[3] = static_cast<float>(orientation[3]);
>  
>    out_position = ApplyNeckModel(out_orientation, 1.0);
>  }
> diff --git a/sdk/include/cardboard.h b/sdk/include/cardboard.h
> index 74a0a67..ed00438 100644
> --- a/sdk/include/cardboard.h
> +++ b/sdk/include/cardboard.h
> @@ -383,6 +383,13 @@ void CardboardHeadTracker_resume(CardboardHeadTracker* head_tracker);
>  
>  /// Gets the predicted head pose for a given timestamp.
>  ///
> +/// @details On Android devices, @p timestamp_ns must be in system boot time
> +///          (see CLOCK_BOOTTIME in https://linux.die.net/man/2/clock_gettime)
> +///          clock (see https://developer.android.com/reference/android/hardware/SensorEvent#timestamp).
> +///          On iOS devices, @p timestamp_ns must be in system uptime raw
> +///          (see CLOCK_UPTIME_RAW in http://www.manpagez.com/man/3/clock_gettime/)
> +///          clock (see https://developer.apple.com/documentation/coremotion/cmlogitem/1615939-timestamp?language=objc).
> +///
>  /// @pre @p head_tracker Must not be null.
>  /// @pre @p position Must not be null.
>  /// @pre @p orientation Must not be null.
> @@ -391,7 +398,7 @@ void CardboardHeadTracker_resume(CardboardHeadTracker* head_tracker);
>  ///
>  /// @param[in]      head_tracker            Head tracker object pointer.
>  /// @param[in]      timestamp_ns            The timestamp for the pose in
> -///     nanoseconds in system monotonic clock.
> +///                                         nanoseconds.
>  /// @param[out]     position                3 floats for (x, y, z).
>  /// @param[out]     orientation             4 floats for quaternion
>  void CardboardHeadTracker_getPose(CardboardHeadTracker* head_tracker,
> diff --git a/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc b/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
> index 0cee973..d5af217 100644
> --- a/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
> +++ b/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc
> @@ -271,7 +271,7 @@ class CardboardApi::CardboardApiImpl {
>        return;
>      }
>      CardboardHeadTracker_getPose(head_tracker_.get(),
> -                                 CardboardApiImpl::GetMonotonicTimeNano() +
> +                                 CardboardApiImpl::GetBootTimeNano() +
>                                       kPredictionTimeWithoutVsyncNanos,
>                                   position, orientation);
>    }
> @@ -476,11 +476,15 @@ class CardboardApi::CardboardApiImpl {
>      }
>    };
>  
> -  // @brief Computes the monotonic time in nano seconds.
> -  // @return The monotonic time count in nano seconds.
> -  static int64_t GetMonotonicTimeNano() {
> +  // @brief Computes the system boot time in nanoseconds.
> +  // @return The system boot time count in nanoseconds.
> +  static int64_t GetBootTimeNano() {
>      struct timespec res;
> -    clock_gettime(CLOCK_MONOTONIC, &res);
> +#if defined(__ANDROID__)
> +    clock_gettime(CLOCK_BOOTTIME, &res);
> +#elif defined(__APPLE__)
> +    clock_gettime(CLOCK_UPTIME_RAW, &res);
> +#endif
>      return (res.tv_sec * CardboardApi::CardboardApiImpl::kNanosInSeconds) +
>             res.tv_nsec;
>    }
> ```

Do you have a resource that I could follow in order to apply the patch.

**Comment by ArwaAlrazooq** [comment]
How do we apply the patch? Specifically for Unity and iOS apps

**Comment by ArwaAlrazooq** [comment]
> How do we apply the patch? Specifically for Unity and iOS apps

I have applied tha patch, tested on Iphone 

> How do we apply the patch? Specifically for Unity and iOS apps

Thanks @jballoffet for the patch, I tested it on iPhone 11, iOS 14.

I'm viewing the sample scene and I can see a better result compare to the current sdk. However it is still laggy. Not as smooth as the gvr. 

Is this the best result we can get now? Has anyone tried the patch and got a smooth movements?

**Comment by maartenslaa** [comment]
Are you testing on gvr on the same device? Some devices have a VrMode to make it feel more responsive. The open source sdk does not support that mode. iOS devices don’t have that. And this may be redundant but make sure the target frame rate is set to 60hz if you’re using unity.

After applying this patch I felt the image could become a little shaky. I added a low pass filter on the velocity to make it less shaky: https://github.com/googlevr/cardboard/pull/212/commits/4002be426a3d2ec9145e79b4bb9da42379526236 Maybe it helps to make it more ‘smooth’ I feel it does.

**Comment by hyenachi** [comment]
Head Tracking of GVR v1.20 with daydream is very good. I think Cardboard xr plugin needs to use daydream's module.

**Comment by chaosemer** [comment]
This was fixed in v1.5.0.
